### ISSUE TITLE
retain bit is not maintained over mosquitto bridge

### ISSUE THREAD

**Issue by victorclaessen** [emqtt/emqttd#1293]
#### Comment
I emailed this to contact@emqtt.io, but after initial receipt response no further communication was received, so I thought I'd just record this issue here.

#### Environment

- OS: Debian 9
- Erlang/OTP: R20/9.0
- EMQ: emqttd-2.3-beta.4

#### Description

It looks like the retain bit is not maintained while bridging to mosquitto.

I have emqttd setup so that the internal listener listens on `127.0.0.1:1883`.

I then have a bridge in mosquitto setup like this:

```listener 21883 127.0.0.1
connection emq
address 127.0.0.1:1883
topic building/pisec/# in 1
try_private true
bridge_protocol_version mqttv311
```

I publish first a retained value to mosquitto:

```mosquitto_pub -t "building/pisec/FX00.07" -m "open" -p 21883 -r```

Then I listen to mosquitto:

```mosquitto_sub -t "building/pisec/FX00.07" -p 21883 -v```

And I publish to emqttd like this:

```mosquitto_pub -t "building/pisec/FX00.07" -m "closed" -r```

And out of the listener I see what I expect:

```building/pisec/FX00.07 closed```

But when I run the listener again to see the retained message

```mosquitto_sub -t "building/pisec/FX00.07" -p 21883 -v -C 1```

Then I get the previous retained value (that I sent directly to mosquitto before)

```building/pisec/FX00.07 open```

So, in other words: The bridge between mosquitto and emqttd is working, but the retained bit is lost in the process. I believe this is what mosquitto uses the "try_private" flag for. I have been able to bridge mosquitto to mosquitto and then the retain bit is carried to the other broker. 

So my thought is that emqttd does not forward the retain bit. Is that true?

**Comment by victorclaessen** [comment]
I see that 2.3-rc.2 has been released. Has this been fixed?

**Comment by emqplus** [comment]
@victorclaessen, The MQTT v3.1.1 specification on this issue:

```
It MUST set the RETAIN flag to 0 when a PUBLISH Packet is sent to a Client because it matches an established subscription regardless of how the flag was set in the message it received [MQTT-3.3.1-9].
```

**Comment by victorclaessen** [comment]
Well, then I suppose we'd better call this a feature request.

From the mqtt.org wiki at https://github.com/mqtt/mqtt.github.io/wiki/bridge_protocol : 

```Bridges are essentially just client connections to another broker.

RSMB added (and Mosquitto followed) a non-spec feature that allows an RSMB
to recognise an inbound connection as another RSMB bridge. From a thread on the google group:

You are right, RSMB recognizes when another RSMB connects to it via a bridge
by using a different protocol version number (when try_private is true, which is the default).
If that connect fails, then the bridge will fall back to MQTT 3.1. This internal protocol is
the same as MQTT 3.1 except that:

        the retained flags are propagated on publish packets, 
        and the subscriptions are nolocal (in the JMS sense), 
        to reduce the chances of infinite message loops.

This behaviour is easy to apply when the broker is the one making the bridge (outgoing topics)
because it knows that the client is a bridge. For incoming topics the different protocol 
number is used. The protocol number is the actual protocol number (3) OR'd with 128 (bit 8).

Both RSMB and Mosquitto implement this behaviour - but it is not a formal part
of the protocol specification.```

**Comment by emqplus** [comment]
@victorclaessen Plan to implement the feature in version 2.3.1, which will be released next weekend;)

**Comment by victorclaessen** [comment]
That is awesome!

**Comment by emqplus** [comment]
@victorclaessen The issue has been fixed in `develop` branch: 
https://github.com/emqtt/emqttd/commit/26fb809dbed287854cee2d97fb717cb72751649c

Build EMQ from source of `develop` branch:

```
git clone -b develop https://github.com/emqtt/emq-relx.git 
cd emq-rel && make run
```

**Comment by victorclaessen** [comment]
That is great. Can't wait for the next release!

So I took a fresh Debian 9 VM, and did:
```
apt-get install erlang
git clone -b develop https://github.com/emqtt/emq-relx.git 
cd emq-relx && make run
```
The (last part of) the output is:
```
===> Starting relx build process ...
===> Resolving OTP Applications from directories:
          /opt/emq-relx/deps
          /usr/lib/erlang/lib
          /opt/emq-relx/apps
===> Resolved emqttd-2.3.1
===> Including Erts from /usr/lib/erlang
===> release successfully created!
Exec: /opt/emq-relx/_rel/emqttd/erts-8.2.1/bin/erlexec -boot /opt/emq-relx/_rel/emqttd/releases/2.3.1/emqttd -mode embedded -boot_var ERTS_LIB_DIR /opt/emq-relx/_rel/emqttd/erts-8.2.1/../lib -mnesia dir "/opt/emq-relx/_rel/emqttd/data/mnesia/emq@127.0.0.1" -config /opt/emq-relx/_rel/emqttd/data/configs/app.2017.11.29.13.17.49.config -args_file /opt/emq-relx/_rel/emqttd/data/configs/vm.2017.11.29.13.17.49.args -vm_args /opt/emq-relx/_rel/emqttd/data/configs/vm.2017.11.29.13.17.49.args -- console
Root: /opt/emq-relx/_rel/emqttd
/opt/emq-relx/_rel/emqttd
Protocol 'inet_tcp': the name emq@127.0.0.1 seems to be in use by another Erlang node
erlang.mk:2499: recipe for target 'run' failed
make: *** [run] Error 1
```

And if I try:
```
cd bin
./emqttd console
```
I get:
```
./emqttd: 20: ./emqttd: runner_log_dir: not found
```

**Comment by emqplus** [comment]
@ victorclaessen First `ps -ef | grep beam` to check if there is already a running EMQ. Kill it and then `make run` or `make && cd _rel/emqttd && ./bin/emqttd console`:)

**Comment by victorclaessen** [comment]
My apologies, it appears this VM was not as clean I thought. I cleared the existing beam process, and then started with ```make run```.

I now run into the following problem. When I publish something (directly) to emqtt broker, the message is only accepted if the retain flag is set.
```
root@n014085:~# mosquitto_sub -t "building/#" -v &
root@n014085:~# mosquitto_pub -t "building/pisec/FX00.00" -m "closed"
root@n014085:~# mosquitto_pub -t "building/pisec/FX00.00" -m "closed" -r
building/pisec/FX00.00 closed
```
The output of emqttd in the console is this:
```
=INFO REPORT==== 29-Nov-2017::15:04:44 ===
    alarm_handler: {set,{system_memory_high_watermark,[]}}
starting emqttd on node 'emq@127.0.0.1'
emqttd ctl is starting...[ok]
emqttd hook is starting...[ok]
emqttd router is starting...[ok]
emqttd pubsub is starting...[ok]
emqttd stats is starting...[ok]
emqttd metrics is starting...[ok]
emqttd pooler is starting...[ok]
emqttd trace is starting...[ok]
emqttd client manager is starting...[ok]
emqttd session manager is starting...[ok]
emqttd session supervisor is starting...[ok]
emqttd wsclient supervisor is starting...[ok]
emqttd broker is starting...[ok]
emqttd alarm is starting...[ok]
emqttd mod supervisor is starting...[ok]
emqttd bridge supervisor is starting...[ok]
emqttd access control is starting...[ok]
emqttd system monitor is starting...[ok]
emqttd 2.3.0 is running now
Eshell V8.2.1  (abort with ^G)
(emq@127.0.0.1)1> Load emq_mod_presence module successfully.
dashboard:http listen on 0.0.0.0:18083 with 2 acceptors.
mqtt:tcp listen on 127.0.0.1:11883 with 16 acceptors.
mqtt:tcp listen on 0.0.0.0:1883 with 16 acceptors.
mqtt:ws listen on 0.0.0.0:8083 with 4 acceptors.
mqtt:ssl listen on 0.0.0.0:8883 with 16 acceptors.
mqtt:wss listen on 0.0.0.0:8084 with 4 acceptors.
mqtt:api listen on 0.0.0.0:8080 with 4 acceptors.
15:05:18.498 [error] gen_server <0.1458.0> terminated with reason: no function clause matching emqttd_protocol:clean_retain(false, {mqtt_message,<<0,5,95,31,160,29,251,167,110,124,0,0,5,185,0,1>>,undefined,{<<"mosqpub/5098-n0140...">>,...},...}) line 553
15:05:18.499 [error] CRASH REPORT Process <0.1458.0> with 0 neighbours exited with reason: no function clause matching emqttd_protocol:clean_retain(false, {mqtt_message,<<0,5,95,31,160,29,251,167,110,124,0,0,5,185,0,1>>,undefined,{<<"mosqpub/5098-n0140...">>,...},...}) line 553 in gen_server2:terminate/3 line 1157
15:05:18.499 [error] gen_server <0.1457.0> terminated with reason: no function clause matching emqttd_protocol:clean_retain(false, {mqtt_message,<<0,5,95,31,160,29,251,167,110,124,0,0,5,185,0,1>>,undefined,{<<"mosqpub/5098-n0140...">>,...},...}) line 553
15:05:18.499 [error] CRASH REPORT Process <0.1457.0> with 0 neighbours exited with reason: no function clause matching emqttd_protocol:clean_retain(false, {mqtt_message,<<0,5,95,31,160,29,251,167,110,124,0,0,5,185,0,1>>,undefined,{<<"mosqpub/5098-n0140...">>,...},...}) line 553 in gen_server2:terminate/3 line 1157
15:05:18.500 [error] gen_server <0.1459.0> terminated with reason: no function clause matching emqttd_protocol:clean_retain(false, {mqtt_message,<<0,5,95,31,160,29,251,167,110,124,0,0,5,185,0,1>>,undefined,{<<"mosqpub/5098-n0140...">>,...},...}) line 553
15:05:18.500 [error] Supervisor 'esockd_connection_sup - <0.1394.0>' had child connection started with emqttd_client:start_link([{max_publish_rate,0},{client_idle_timeout,30000},{client_enable_stats,false},{max_clientid_len,...},...]) at <0.1458.0> exit with reason no function clause matching emqttd_protocol:clean_retain(false, {mqtt_message,<<0,5,95,31,160,29,251,167,110,124,0,0,5,185,0,1>>,undefined,{<<"mosqpub/5098-n0140...">>,...},...}) line 553 in context connection_crashed
15:05:18.501 [error] Supervisor 'esockd_connection_sup - <0.1394.0>' had child connection started with emqttd_client:start_link([{max_publish_rate,0},{client_idle_timeout,30000},{client_enable_stats,false},{max_clientid_len,...},...]) at <0.1457.0> exit with reason no function clause matching emqttd_protocol:clean_retain(false, {mqtt_message,<<0,5,95,31,160,29,251,167,110,124,0,0,5,185,0,1>>,undefined,{<<"mosqpub/5098-n0140...">>,...},...}) line 553 in context connection_crashed
15:05:18.501 [error] CRASH REPORT Process <0.1459.0> with 0 neighbours exited with reason: no function clause matching emqttd_protocol:clean_retain(false, {mqtt_message,<<0,5,95,31,160,29,251,167,110,124,0,0,5,185,0,1>>,undefined,{<<"mosqpub/5098-n0140...">>,...},...}) line 553 in gen_server2:terminate/3 line 1157
15:05:18.501 [error] Supervisor 'esockd_connection_sup - <0.1394.0>' had child connection started with emqttd_client:start_link([{max_publish_rate,0},{client_idle_timeout,30000},{client_enable_stats,false},{max_clientid_len,...},...]) at <0.1459.0> exit with reason no function clause matching emqttd_protocol:clean_retain(false, {mqtt_message,<<0,5,95,31,160,29,251,167,110,124,0,0,5,185,0,1>>,undefined,{<<"mosqpub/5098-n0140...">>,...},...}) line 553 in context connection_crashed
```
That 15:05:18.498 crash report was when I published without the retain flag (and _before_ I published the same message _with_ the retain flag).

**Comment by emqplus** [comment]
@victorclaessen `cd deps/emqttd && git pull` and then `make run` again. 

@huangdan Please add unit test cases for the clean_retain/2 function.

**Comment by victorclaessen** [comment]
Ok, now the crash does not occur anymore, and the message is propagated.

Also, I can confirm that the retain flag is carried across the bridge if (and _only_ if) _try_private_ is set to _true_ in the mosquitto configuration.

Thanks a lot, this makes bridging much more flexible and useful!
